File: backend/streaming/entity_parser.py
"""Entity extraction from markdown markers.

The agent emits structured entities using markdown markers:
- :::contact ... ::: for Contact cards
- :::event ... ::: for CalendarEvent cards

Entities are embedded in JSON code blocks within the markers.
This parser extracts and validates them against the frontend schemas.
"""
import json
import re
import logging
from typing import Literal, TypedDict, Optional
from enum import Enum

logger = logging.getLogger(__name__)


class EntityType(str, Enum):
    CONTACT = "contact"
    EVENT = "event"


class ContactEntity(TypedDict, total=False):
    """Matches frontend ContactCardProps interface."""
    name: str  # required
    email: Optional[str]
    phone: Optional[str]
    address: Optional[str]
    avatar: Optional[str]


class CalendarEventEntity(TypedDict, total=False):
    """Matches frontend CalendarEventProps interface."""
    title: str  # required
    date: str   # required
    startTime: Optional[str]
    endTime: Optional[str]
    location: Optional[str]
    description: Optional[str]


class ExtractedEntity(TypedDict):
    """Extracted entity with type and data."""
    type: EntityType
    data: ContactEntity | CalendarEventEntity
    raw: str  # Original marker text for replacement


# Regex patterns for entity markers
# Format: :::type\n```json\n{...}\n```\n:::
CONTACT_PATTERN = re.compile(
    r':::contact\s*\n```json\s*\n(.*?)\n```\s*\n:::',
    re.DOTALL
)
EVENT_PATTERN = re.compile(
    r':::event\s*\n```json\s*\n(.*?)\n```\s*\n:::',
    re.DOTALL
)


def extract_entities(markdown: str) -> list[ExtractedEntity]:
    """Extract all entities from markdown content.

    Args:
        markdown: The markdown content to parse

    Returns:
        List of extracted entities with type, data, and raw marker text
    """
    entities = []

    # Extract contacts
    for match in CONTACT_PATTERN.finditer(markdown):
        try:
            json_str = match.group(1).strip()
            data = json.loads(json_str)

            # Validate required field
            if "name" not in data:
                logger.warning(f"Contact missing required 'name' field: {json_str[:50]}")
                continue

            entities.append({
                "type": EntityType.CONTACT,
                "data": data,
                "raw": match.group(0),
            })
        except json.JSONDecodeError as e:
            logger.warning(f"Failed to parse contact JSON: {e}")
            continue

    # Extract events
    for match in EVENT_PATTERN.finditer(markdown):
        try:
            json_str = match.group(1).strip()
            data = json.loads(json_str)

            # Validate required fields
            if "title" not in data or "date" not in data:
                logger.warning(f"Event missing required fields: {json_str[:50]}")
                continue

            entities.append({
                "type": EntityType.EVENT,
                "data": data,
                "raw": match.group(0),
            })
        except json.JSONDecodeError as e:
            logger.warning(f"Failed to parse event JSON: {e}")
            continue

    return entities


def validate_contact(data: dict) -> bool:
    """Validate contact data against ContactCardProps schema."""
    if not isinstance(data.get("name"), str):
        return False
    # Optional fields can be missing or string
    for field in ["email", "phone", "address", "avatar"]:
        if field in data and not isinstance(data[field], str):
            return False
    return True


def validate_event(data: dict) -> bool:
    """Validate event data against CalendarEventProps schema."""
    if not isinstance(data.get("title"), str):
        return False
    if not isinstance(data.get("date"), str):
        return False
    # Optional fields
    for field in ["startTime", "endTime", "location", "description"]:
        if field in data and not isinstance(data[field], str):
            return False
    return True

File: frontend/lib/entity-parser.ts
/**
 * Entity parser for extracting Contact and CalendarEvent from markdown markers.
 *
 * The agent emits structured entities using markdown markers:
 * - :::contact\n```json\n{...}\n```\n::: for Contact cards
 * - :::event\n```json\n{...}\n```\n::: for CalendarEvent cards
 *
 * This parser extracts these markers and returns structured data
 * that EntityRenderer can display.
 */

import type { ContactCardProps, CalendarEventProps } from '@/types';

export type EntityType = 'contact' | 'event';

export interface ParsedContact {
  type: 'contact';
  data: ContactCardProps;
  raw: string;
  startIndex: number;
  endIndex: number;
}

export interface ParsedEvent {
  type: 'event';
  data: CalendarEventProps;
  raw: string;
  startIndex: number;
  endIndex: number;
}

export type ParsedEntity = ParsedContact | ParsedEvent;

export interface ParseResult {
  /** Text segments between entities */
  textSegments: string[];
  /** Parsed entities in order of appearance */
  entities: ParsedEntity[];
  /** Whether content has incomplete entity marker (still streaming) */
  hasIncompleteEntity: boolean;
}

// Regex patterns matching backend format
// Format: :::type\n```json\n{...}\n```\n:::
const CONTACT_PATTERN = /:::contact\s*\n```json\s*\n([\s\S]*?)\n```\s*\n:::/g;
const EVENT_PATTERN = /:::event\s*\n```json\s*\n([\s\S]*?)\n```\s*\n:::/g;

// Pattern to detect incomplete entity markers (still streaming)
const INCOMPLETE_PATTERN = /:::(?:contact|event)(?:\s*\n```json\s*\n[\s\S]*)?$/;

/**
 * Parse entities from markdown content.
 *
 * Extracts Contact and CalendarEvent entities from markdown markers,
 * returning both the entities and the remaining text segments.
 *
 * @param content - The markdown content to parse
 * @returns ParseResult with text segments, entities, and streaming status
 */
export function parseEntities(content: string): ParseResult {
  const entities: ParsedEntity[] = [];
  const allMatches: Array<{ match: RegExpExecArray; type: EntityType }> = [];

  // Find all contact matches
  let match: RegExpExecArray | null;
  const contactRegex = new RegExp(CONTACT_PATTERN.source, 'g');
  while ((match = contactRegex.exec(content)) !== null) {
    allMatches.push({ match, type: 'contact' });
  }

  // Find all event matches
  const eventRegex = new RegExp(EVENT_PATTERN.source, 'g');
  while ((match = eventRegex.exec(content)) !== null) {
    allMatches.push({ match, type: 'event' });
  }

  // Sort by position in content
  allMatches.sort((a, b) => a.match.index - b.match.index);

  // Parse each match
  for (const { match, type } of allMatches) {
    try {
      const jsonStr = match[1].trim();
      const data = JSON.parse(jsonStr);

      if (type === 'contact') {
        // Validate required field
        if (typeof data.name !== 'string') {
          console.warn('Contact missing required name field:', jsonStr.slice(0, 50));
          continue;
        }
        entities.push({
          type: 'contact',
          data: data as ContactCardProps,
          raw: match[0],
          startIndex: match.index,
          endIndex: match.index + match[0].length,
        });
      } else if (type === 'event') {
        // Validate required fields
        if (typeof data.title !== 'string' || typeof data.date !== 'string') {
          console.warn('Event missing required fields:', jsonStr.slice(0, 50));
          continue;
        }
        entities.push({
          type: 'event',
          data: data as CalendarEventProps,
          raw: match[0],
          startIndex: match.index,
          endIndex: match.index + match[0].length,
        });
      }
    } catch (e) {
      console.warn(`Failed to parse ${type} JSON:`, e);
      continue;
    }
  }

  // Build text segments (content between entities)
  const textSegments: string[] = [];
  let lastEnd = 0;

  for (const entity of entities) {
    // Text before this entity
    if (entity.startIndex > lastEnd) {
      textSegments.push(content.slice(lastEnd, entity.startIndex));
    }
    lastEnd = entity.endIndex;
  }

  // Text after last entity
  if (lastEnd < content.length) {
    const remaining = content.slice(lastEnd);
    // Check if remaining has incomplete entity marker
    const hasIncomplete = INCOMPLETE_PATTERN.test(remaining);
    if (hasIncomplete) {
      // Don't include incomplete marker in text segments
      const incompleteMatch = remaining.match(INCOMPLETE_PATTERN);
      if (incompleteMatch && incompleteMatch.index !== undefined) {
        textSegments.push(remaining.slice(0, incompleteMatch.index));
      }
    } else {
      textSegments.push(remaining);
    }
  }

  return {
    textSegments,
    entities,
    hasIncompleteEntity: INCOMPLETE_PATTERN.test(content.slice(lastEnd)),
  };
}

/**
 * Check if content contains any entity markers (complete or incomplete).
 */
export function hasEntityMarkers(content: string): boolean {
  return content.includes(':::contact') || content.includes(':::event');
}
